pkg/cleanup/users.go

```diff
@@ -16,28 +16,12 @@ func (u *users) Name() string {
 	return "remove k0s users step:"
 }
 
-// NeedsToRun detects controller users
-func (u *users) NeedsToRun() bool {
-	clusterConfig, err := config.GetYamlFromFile(u.Config.cfgFile, u.Config.k0sVars)
-	if err != nil {
-		return false
-	}
-
-	users := install.GetControllerUsers(clusterConfig)
-	for _, user := range users {
-		if exists, _ := iusers.CheckIfUserExists(user); exists {
-			return true
-		}
-	}
-	return false
-}
-
 // Run removes all controller users that are present on the host
 func (u *users) Run() error {
 	logger := logrus.New()
 	clusterConfig, err := config.GetYamlFromFile(u.Config.cfgFile, u.Config.k0sVars)
 	if err != nil {
-		logger.Errorf("failed to get cluster setup: %v", err)
+		return fmt.Errorf("failed to get cluster setup: %v", err)
```

Comment:
This changes the behavior, the function never returned an error even if some parts of it failed.

Reply:
The `clusterConfig` (`cfg` after recent rebase) value is undefined if there are errors and is likely nil, in which case it would segfault.

`install.DeleteControllerUsers(nil)` ends with a segfault.
```
INFO[2021-09-29 12:13:36] * containers steps                           
INFO[2021-09-29 12:13:36] * remove k0s users step:                     
INFO[2021-09-29 12:13:36] no config file given, using defaults         
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x128 pc=0x1b5e57a]

goroutine 1 [running]:
github.com/k0sproject/k0s/pkg/install.DeleteControllerUsers(0x0, 0x0, 0xc0003c7be0)
	/go/src/github.com/k0sproject/k0s/pkg/install/users.go:54 +0x3a
github.com/k0sproject/k0s/pkg/cleanup.(*users).Run(0xc000378870, 0x4, 0xc0005c3250)
	/go/src/github.com/k0sproject/k0s/pkg/cleanup/users.go:27 +0x20a
github.com/k0sproject/k0s/pkg/cleanup.(*Config).Cleanup(0xc000ebe4e0, 0x1b, 0xc0008771a0)
	/go/src/github.com/k0sproject/k0s/pkg/cleanup/cleanup_unix.go:89 +0x2f9
github.com/k0sproject/k0s/cmd/reset.(*CmdOpts).reset(0xc0005c3878, 0x0, 0x33b18a9)
	/go/src/github.com/k0sproject/k0s/cmd/reset/reset.go:74 +0x2ce
github.com/k0sproject/k0s/cmd/reset.NewResetCmd.func1(0xc000caaf00, 0x51d2df8, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/k0sproject/k0s/cmd/reset/reset.go:42 +0xc5
github.com/spf13/cobra.(*Command).execute(0xc000caaf00, 0x51d2df8, 0x0, 0x0, 0xc000caaf00, 0x51d2df8)
	/gocache/mod/github.com/spf13/cobra@v1.2.1/command.go:856 +0x472
github.com/spf13/cobra.(*Command).ExecuteC(0xc000692c80, 0x407905, 0xc00010e058, 0x2b09990)
	/gocache/mod/github.com/spf13/cobra@v1.2.1/command.go:974 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
	/gocache/mod/github.com/spf13/cobra@v1.2.1/command.go:902
github.com/k0sproject/k0s/cmd.Execute()
	/go/src/github.com/k0sproject/k0s/cmd/root.go:206 +0x2a
main.main()
	/go/src/github.com/k0sproject/k0s/main.go:40 +0x25
```
So this is technically a bug fix.

I guess that should be mentioned in the commit message.

Reply:
Oh, now I see what you mean. It will result in `k0s reset` returns error. I guess if it fails to parse the config it should return failure, shouldn't it?

Reply:
I'm not sure. Maybe it should just skip trying to do whatever is based on configuration and just do what it can outside of it (remove files, uninstall services, whatever?)